gapis/api/vulkan/frame_loop.go

```diff
@@ -709,3 +798,109 @@ func (f *frameLoop) resetDescriptorSetLayouts(ctx context.Context, stateBuilder
 
 	return nil
 }
+
+func (f *frameLoop) resetCommandBuffers(ctx context.Context, stateBuilder *stateBuilder) error {
+
+	for cmdBuf := range f.commandBufferAllocated {
+		log.D(ctx, "Command buffer %v allocated during loop, free it.", cmdBuf)
+		cmdBufObj := GetState(f.loopEndState).CommandBuffers().Get(cmdBuf)
+		if cmdBufObj != NilCommandBufferObjectʳ {
+			stateBuilder.write(stateBuilder.cb.VkFreeCommandBuffers(
+				cmdBufObj.Device(),
+				cmdBufObj.Pool(),
+				1,
+				stateBuilder.MustAllocReadData(cmdBufObj.VulkanHandle()).Ptr(),
+			))
+		} else {
+			log.E(ctx, "Command buffer %v cannot be found in loop ending state", cmdBuf)
+		}
+	}
+
+	for cmdBuf := range f.commandBufferFreed {
+		cmdBufObj := GetState(f.loopStartState).CommandBuffers().Get(cmdBuf)
+		if cmdBufObj == NilCommandBufferObjectʳ {
+			log.E(ctx, "Command buffer %v can not be found in loop starting state", cmdBuf)
+			continue
+		}
+		log.D(ctx, "Command buffer %v freed during loop, recreate it.", cmdBuf)
+		// sb.createCommandBuffer(cmdBufObj, cmdBufObj.Level())
+		stateBuilder.write(stateBuilder.cb.VkAllocateCommandBuffers(
+			cmdBufObj.Device(),
+			stateBuilder.MustAllocReadData(NewVkCommandBufferAllocateInfo(stateBuilder.ta,
+				VkStructureType_VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO, // sType
+				0,                 // pNext
+				cmdBufObj.Pool(),  // commandPool
+				cmdBufObj.Level(), // level
+				1,                 // commandBufferCount
+			)).Ptr(),
+			stateBuilder.MustAllocWriteData(cmdBufObj.VulkanHandle()).Ptr(),
+			VkResult_VK_SUCCESS,
+		))
+
+		f.reRecordCommandBuffer(ctx, stateBuilder, cmdBufObj, cmdBuf)
+	}
+
+	for cmdBuf := range f.commandBufferChanged {
+		cmdBufObj := GetState(f.loopStartState).CommandBuffers().Get(cmdBuf)
+		if cmdBufObj == NilCommandBufferObjectʳ {
+			log.E(ctx, "Command buffer %v can not be found in loop starting state", cmdBuf)
+		}
+		log.D(ctx, "Command buffer %v changed during loop, re-record it.", cmdBuf)
+		if err := f.reRecordCommandBuffer(ctx, stateBuilder, cmdBufObj, cmdBuf); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f *frameLoop) reRecordCommandBuffer(ctx context.Context, stateBuilder *stateBuilder, refCmdBuf CommandBufferObjectʳ, commandBufferID VkCommandBuffer) error {
```

Comment:
Is there nothing in StateBuilder.go that can help here? It seems like something that would be required by MEC?